pkg/action/install.go

```diff
@@ -584,14 +592,107 @@ func (c *Configuration) renderResources(ch *chart.Chart, values chartutil.Values
 		}
 	}
 
-	if pr != nil {
-		b, err = pr.Run(b)
+	return hs, b, notes, nil
+}
+
+// postProcessFiles sends files for post-processing.
+// Some post-processing may need all objects at once.
+// Therefore we concatenate all rendered files, post-process and split them back.
+// A temporary annotation is added to preserve template filename.
+func (c *Configuration) postProcessFiles(files map[string]string, pr postrender.PostRenderer) (map[string]string, error) {
+	if pr == nil {
+		return files, nil
+	}
+
+	annotationName := "helmPostProcessFileName"
+
+	b := bytes.NewBuffer(nil)
+
+	// Split the rendered files into documents and add the temp annotation.
+	reYamlDocumentSeparator := regexp.MustCompile(`\n---\s*\n`)
+	for fileName, fileContent := range files {
+		for _, document := range reYamlDocumentSeparator.Split(fileContent, -1) {
+			if strings.TrimSpace(document) == "" {
+				continue
+			}
+			b.WriteString("---\n")
+			documentYaml, _, err := yamlSetTemporaryAnnotation(document, annotationName, fileName)
```

Comment:
This is an interesting approach. My one concern about this is what happens if the post-render step mucks with the annotations (either removing unknown annotations or setting their own). Are there any other ways we could do this without an annotation?

Reply:
The options are:
1. Run the post-renderer separately for non-hook templates (`[]byte`) and for each hook.

2. Remember template filename string for each object that is being marshalled for post-processing. For example, before postprocessing, fill the `map[string]string`, where `key:=fmt.Sprintf("%s;%s;$s", apiVersion, kind, metaName)` and value is the filename. Upon unmarshalling, use that map to find the filename. If not found, assign some random string.

3. Use [any other field from metadata](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta), such as name or label. Unfortunately, other fields have more strict limitations on length and a set of allowed characters. IMHO, `annotations` is the best option there.

4. Use annotations, but if our `helm.sh/separator` annotation is not found, then "adopt" such object with a random file name. Are the original template filenames important after all? Don't know how they usedapart from a comment in the final yaml.

5. Mix options 2 and 4 by running post-renderer twice. Use first run with extra annotation to obtain extra mapping. In second run, post-render without extra annotations and assign a filename which best matches both mappings.

Option 4 is implemented now.

Option 5 may be dangerous if a user's postrenderer script is not idempotent.

I personally like the second option too. `kind` almost always stays the same. `apiVersion` often gets changed to a similar value. `metadata.name` is usually prefixed or suffixed with a string. So it may worth a try. I just don't know (yet) how to implemement a string similarity algorithm which works good for `metadata.name` and especially for `apiVersion` comparison.